File: icu4j/main/classes/core/src/com/ibm/icu/text/ScientificFormatHelper.java
/*
 *******************************************************************************
 * Copyright (C) 2014, International Business Machines Corporation and
 * others. All Rights Reserved.
 *******************************************************************************
 */
package com.ibm.icu.text;

import java.text.AttributedCharacterIterator;

/**
 * A helper class for formatting in user-friendly scientific notation.
 * 
 * ScientificFormatHelper instances are immutable and thread-safe. However, the
 * AttributedCharacterIterator instances that ScientificFormatHelper instances format must
 * not be shared across multiple threads.
 *
 * Sample code:
 * <pre>
 * DecimalFormat decfmt = (DecimalFormat) NumberFormat.getScientificInstance(new ULocale("en"));
 * AttributedCharacterIterator iterator = decfmt.formatToCharacterIterator(1.23456e-78);
 * ScientificFormatHelper helper = ScientificFormatHelper.getInstance(
                decfmt.getDecimalFormatSymbols());
 * <pre>
 * // Output: "1.23456×10<sup>-78</sup>"
 *  System.out.println(helper.insertMarkup(iterator, "<sup>", "</sup>"));
 * </pre>
 *
 * @see NumberFormat
 * @draft ICU 54
 * @provisional This API might change or be removed in a future release.
 *
 */
public final class ScientificFormatHelper {
    
    private final String preExponent;
    
    private ScientificFormatHelper(String preExponent) {
        this.preExponent = preExponent;
    }

    /**
     * Returns a new ScientificFormatHelper.
     * @param dfs comes from the DecimalFormat instance used for default scientific notation.
     * @draft ICU 54
     * @provisional This API might change or be removed in a future release.
     */
    public static ScientificFormatHelper getInstance(DecimalFormatSymbols dfs) {
        return new ScientificFormatHelper(ScientificNumberFormatter.getPreExponent(dfs));
    }

    /**
     * Makes scientific notation user-friendly by surrounding exponent with
     * html to make it superscript.
     * @param iterator the value that DecimalFormat.formatToCharacterIterator() returned.
     * @param beginMarkup the start html for the exponent e.g "<sup>"
     * @param endMarkup the end html for the exponent e.g "</sup>"
     * @return the user-friendly scientific notation.
     * @draft ICU 54
     * @provisional This API might change or be removed in a future release.
     */
    public String insertMarkup(
            AttributedCharacterIterator iterator,
            CharSequence beginMarkup,
            CharSequence endMarkup) {
        return format(
                iterator,
                new ScientificNumberFormatter.MarkupStyle(
                        beginMarkup.toString(), endMarkup.toString()));
    }

    /**
     * Makes scientific notation user-friendly by using specific code points
     * for superscript 0..9, -, and + in the exponent rather than by using
     * html.
     * @param iterator the value that DecimalFormat.formatToCharacterIterator() returned.
     * @return the user-friendly scientific notation.
     * @draft ICU 54
     * @provisional This API might change or be removed in a future release.
     */
    public String toSuperscriptExponentDigits(AttributedCharacterIterator iterator) {
        return format(iterator, ScientificNumberFormatter.SUPER_SCRIPT);
    }
    
    private String format(
            AttributedCharacterIterator iterator, ScientificNumberFormatter.Style option) {
        return option.format(iterator, preExponent);
    }
}

File: source/i18n/scientificformathelper.cpp
/*
**********************************************************************
* Copyright (c) 2014, International Business Machines
* Corporation and others.  All Rights Reserved.
**********************************************************************
*/
#include "unicode/utypes.h"

#if !UCONFIG_NO_FORMATTING

#include "unicode/scientificformathelper.h"
#include "unicode/scientificnumberformatter.h"
#include "decfmtst.h"

U_NAMESPACE_BEGIN

ScientificFormatHelper::ScientificFormatHelper(
        const DecimalFormatSymbols &dfs, UErrorCode &status)
        : fPreExponent(), fStaticSets(NULL) {
    if (U_FAILURE(status)) {
        return;
    }
    ScientificNumberFormatter::getPreExponent(dfs, fPreExponent);
    fStaticSets = DecimalFormatStaticSets::getStaticSets(status);
}

ScientificFormatHelper::ScientificFormatHelper(
        const ScientificFormatHelper &other)
        : UObject(other),
          fPreExponent(other.fPreExponent),
          fStaticSets(other.fStaticSets) {
}

ScientificFormatHelper &ScientificFormatHelper::operator=(const ScientificFormatHelper &other) {
    if (this == &other) {
        return *this;
    }
    fPreExponent = other.fPreExponent;
    fStaticSets = other.fStaticSets;
    return *this;
}

ScientificFormatHelper::~ScientificFormatHelper() {
}

UnicodeString &ScientificFormatHelper::insertMarkup(
        const UnicodeString &s,
        FieldPositionIterator &fpi,
        const UnicodeString &beginMarkup,
        const UnicodeString &endMarkup,
        UnicodeString &result,
        UErrorCode &status) const {
    if (U_FAILURE(status)) {
        return result;
    }
    ScientificNumberFormatter::MarkupStyle style(beginMarkup, endMarkup);
    return style.format(
            s,
            fpi,
            fPreExponent,
            *fStaticSets,
            result,
            status);
}

UnicodeString &ScientificFormatHelper::toSuperscriptExponentDigits(
        const UnicodeString &s,
        FieldPositionIterator &fpi,
        UnicodeString &result,
        UErrorCode &status) const {
    if (U_FAILURE(status)) {
      return result;
    }
    ScientificNumberFormatter::SuperscriptStyle style;
    return style.format(
            s,
            fpi,
            fPreExponent,
            *fStaticSets,
            result,
            status);
}

U_NAMESPACE_END

#endif /* !UCONFIG_NO_FORMATTING */
